### PR TITLE
Bump libraries, add portForward method

### PR DESCRIPTION
**Pull request by gingerbreadmanphoto**

### CODE REVIEW

**Review by rintcius — APPROVED** [review]

File: core/src/main/scala/fs2/io/ssh/Client.scala [hunk 1]
```diff
@@ -17,65 +17,71 @@
...
           // TODO handle failure opening
         } yield channel)(
         channel => fromFuture(F.delay(channel.close(false))).void)
     } yield new Process[F](channel, chunkSize)
   }
 
+  def portForward(cc: ConnectionConfig,
+                  local: InetSocketAddress,
+                  remote: InetSocketAddress)(
+                   implicit FR: Raise[F, Error]): Resource[F, Unit] = {
```
Comment (wemrysi) [inline core/src/main/scala/fs2/io/ssh/Client.scala]:
```suggestion
  def portForward(
      cc: ConnectionConfig,
      local: InetSocketAddress,
      remote: InetSocketAddress)(
      implicit FR: Raise[F, Error]): Resource[F, Unit] = {
```

File: core/src/main/scala/fs2/io/ssh/Client.scala [hunk 2]
```diff
@@ -17,65 +17,71 @@
 package fs2
 package io.ssh
 
-import cats.effect.{Blocker, Concurrent, ContextShift, Resource, Sync}
+import cats.effect.kernel.Async
```
Comment (wemrysi) [inline core/src/main/scala/fs2/io/ssh/Client.scala]: Prefer omitting `kernel` for consistency, generally
```suggestion
import cats.effect.Async
```

File: core/src/main/scala/fs2/io/ssh/MinaFuture.scala [hunk 3]
```diff
@@ -121,9 +114,7 @@ private[ssh] object MinaFuture {
     def exception(s: IoReadFuture) = s.getException()
 
     def result(s: IoReadFuture) = {
-      val buffer = s.getBuffer()
-      val length = s.getRead()
-      Chunk.bytes(buffer.array(), 0, length)
+      Chunk.array(s.getBuffer().array())
```
Comment (wemrysi) [inline core/src/main/scala/fs2/io/ssh/MinaFuture.scala]: This should be reverted to the previous impl, but using `Chunk.array(Array[Byte], Int, Int)` instead.

The problem with the current implementation is that `Buffer#array` returns the raw underlying `Array[Byte]`, which may be physically larger than the number of bytes read. `IoReadFuture#getRead` tells us how many bytes were actually read, so we know how much of the array it is safe to consume.

File: core/src/main/scala/fs2/io/ssh/Process.scala [hunk 4]
```diff
@@ -86,16 +81,16 @@ final class Process[F[_]: Concurrent: ContextShift] private[ssh] (
   private[this] def ioosToSink(ioosF: F[IoOutputStream]): Pipe[F, Byte, Unit] = { in =>
     Stream.eval(ioosF) flatMap { ioos =>
       val written = in.chunks evalMap { chunk =>
-        val bytes = chunk.toBytes
-        val buffer = new ByteArrayBuffer(bytes.values, bytes.offset, bytes.length)
+        val bytes = chunk.toArray
```
Comment (wemrysi) [inline core/src/main/scala/fs2/io/ssh/Process.scala]: This should also be reverted to the previous implementation, but using `chunk.toArraySlice`. This results in fewer allocations when `Chunk` is already backed by an `Array[Byte]`.

File: core/src/main/scala/fs2/io/ssh/Process.scala [hunk 5]
```diff
@@ -86,16 +81,16 @@ final class Process[F[_]: Concurrent: ContextShift] private[ssh] (
...
+        val buffer = new ByteArrayBuffer(bytes, 0, bytes.length)
         buffer.wpos(bytes.length)
```
Comment (wemrysi) [inline core/src/main/scala/fs2/io/ssh/Process.scala]: This line is no longer necessary (it is actually incorrect and should have been `bytes.offset + bytes.length`) as the [`ByteArrayBuffer(byte[], Int, Int)` constructor](https://github.com/apache/mina-sshd/blob/sshd-2.9.2/sshd-common/src/main/java/org/apache/sshd/common/util/buffer/ByteArrayBuffer.java#L102) already handles this.

**Review by wemrysi — CHANGES_REQUESTED** [review]
Is it feasible to add a test for the port forwarding feature?

File: core/src/main/scala/fs2/io/ssh/MinaFuture.scala (hunk 3 above)
Comment (rintcius) [reply core/src/main/scala/fs2/io/ssh/MinaFuture.scala]: Oh good to know! @gingerbreadmanphoto could you document this?

File: core/src/main/scala/fs2/io/ssh/Process.scala (hunk 4 above)
Comment (rintcius) [reply core/src/main/scala/fs2/io/ssh/Process.scala]: Same here - good to document this too

File: core/src/main/scala/fs2/io/ssh/MinaFuture.scala (hunk 3 above)
Comment (gingerbreadmanphoto) [reply core/src/main/scala/fs2/io/ssh/MinaFuture.scala]: Yeah 👍

File: core/src/main/scala/fs2/io/ssh/Process.scala (hunk 4 above)
Comment (gingerbreadmanphoto) [reply core/src/main/scala/fs2/io/ssh/Process.scala]: 👍

File: core/src/main/scala/fs2/io/ssh/Process.scala (hunk 5 above)
Comment (gingerbreadmanphoto) [reply core/src/main/scala/fs2/io/ssh/Process.scala]: Removed 👍

File: core/src/main/scala/fs2/io/ssh/Client.scala (hunk 2 above)
Comment (gingerbreadmanphoto) [reply core/src/main/scala/fs2/io/ssh/Client.scala]: 👍

File: core/src/main/scala/fs2/io/ssh/Client.scala (hunk 1 above)
Comment (gingerbreadmanphoto) [reply core/src/main/scala/fs2/io/ssh/Client.scala]: 👍

**Comment by gingerbreadmanphoto** [comment]
> Is it feasible to add a test for the port forwarding feature?

Added a test 👍

**Review by wemrysi — APPROVED** [review]

**Pull request merged**